### ISSUE TITLE
Bug: [sort-type-constituents] strange organization of constituents

### ISSUE THREAD

**Issue by rentalhost** [typescript-eslint/typescript-eslint#6787]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.2&sourceType=module&code=MYGwhgzhAEAiYBcCmAxA9gJwLaOgb2gF8BYAKFEhgBUBLLVTHBfIssgenegEkA7ABwCuzSNCQAPfkmDIAJgC4yCAJ5ToAUXFgs-EEgCM0ALxlo0AD5xEDbIlMXoteulsJ7liAgw1eAcwDaALru0J7efmyknNAAQmCyoYK+vkieNGi8iqQqapraukgATMYh8MguTCFhPgGB0NEAPAC0TdAA6gASAJoA-CFONpWkZh5eNWRAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAe2jqNNrI5Nk+SvliIWqDJETRoXSODABfEMqA

### Repro Code

```typescript
class DateFormat { }
class TimeFormat { }

// Input as expected:
type Example1 =
  | DateFormat
  | TimeFormat
  | string[]
  | string

// Bad suggestion:
type Example2 =
  | DateFormat
  | string[] // <-- WHY?
  | TimeFormat
  | string
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/sort-type-constituents": "error"
  }
}
```

### Expected Result

Do not mark as error.

### Actual Result

`string[]` is suggested to put between `DateFormat` (a class type) and `TimeFormat` (another class type).

**Comment by bradzacher** [comment]
This is working as expected. Array types are treated like any other type, not like the type of their element type.

Eg `string[]` is not grouped as a "primitive" - but as a "named" type.

See the example in the rule docs which shows this.
https://typescript-eslint.io/rules/sort-type-constituents#examples
